### PR TITLE
Update the testing guide within the development guide

### PR DESCRIPTION
**Pull request by namurphy**
There are a couple of parts of our testing guide that have become out-of-date.  For example, we no longer use Travis CI or AppVeyor.  This PR is intended to update the testing guide to make it more up-to-date.

This is a companion pull request to #1156.

### CODE REVIEW

**Comment by namurphy** [comment]
I'd like to finish this for 0.7.0 so I tentatively marked that milestone, but it'll more likely be in 0.8.0.

File: docs/development/testing_guide.rst [hunk 1]
```diff
@@ -1,50 +1,49 @@
...
 easier to understand and modify when the behavior of a function or
-method is changed.  Consequently, tests should be held to the same
+method is changed. Consequently, tests should be held to the same
 code quality standards as the rest of the package.
 
-When bugs are discovered, they should be turned into test cases to
-prevent the bug from emerging again in the future [2]_.
+.. tip::
+
+   Writing tests takes time, but debugging takes more time.
```
Comment (namurphy) [inline docs/development/testing_guide.rst:24]: Thanks to @StanczakDominik for this quote from the software testing tutorial during Plasma Hack Week!

Comment (StanczakDominik) [reply docs/development/testing_guide.rst:24]: ... I spent most of yesterday debugging tests. Should we mention that?

File: docs/development/testing_guide.rst [hunk 2]
```diff
@@ -60,31 +59,24 @@ The recommended way to run PlasmaPy's full test suite when contributing
...
-documentation in both HTML and LaTeX formats, and reports any errors
-that arise.
-
-If any inconsistencies with the `PEP 8 style guide
-<https://www.python.org/dev/peps/pep-0008/?>`_ are found, then
-`pep8speaks <https://pep8speaks.com/>`_ will comment on the pull request
-and update that comment as the pull request is updated.
+If any inconsistencies with the :pep:`8` style guide are found, then
+automated code style checks will and update that comment as the pull
+request is updated.  (NEEDS UPDATING)
```
Comment (StanczakDominik) [inline docs/development/testing_guide.rst:79]:
```suggestion
PlasmaPy uses the `black` code formatter. If any inconsistencies
with its style are found, the `pre-commit.ci - pr` will notify you
about them. To apply the fixes automagically, write a comment with
the message `pre-commit.ci autofix`. Remember to `git pull`
afterwards!
```

**Review by StanczakDominik — APPROVED** [review]
This honestly LGTM. Added a single fix to the style check guide.

Also, black is not quite pep8.

**Comment by namurphy** [comment]
So I don't forget from today's community meeting...

 - [ ] Add a subsection on running tests with pytest. Mention that it is faster, is less resource intensive, and has less computational overhead.  Mention that it is recommended for running tests locally
 - [ ] Add a subsection on running tests with tox.  Mention that it's less effort to run since you don't have to set up an environment, but that because it creates a virtual environment, the tests take longer to run.  Mention that it is well-suited for CI because of that.
 - [ ] Remove `python setup.py test`.

File: docs/development/testing_guide.rst [hunk 3]
```diff
@@ -1,219 +1,263 @@
...
+require any extra work. The disadvantages are that running the tests on
+GitHub_ is often slow and that navigating the test results is sometimes
+difficult.
 
-Motivation
-==========
+We recommend that experienced contributors run tests either by using
+pytest_ from the command line or by using your preferred IDE.
+Using tox_ is an alternative to pytest_, but running tests with tox_ is
+typically slower than running tests with pytest_.
```
Comment (StanczakDominik) [inline docs/development/testing_guide.rst:61]:
```suggestion
Using tox_ is an alternative to pytest_, but running tests with tox_ adds
the overhead of creating an isolated environment for your test and can
thus be slower.
```

File: docs/development/testing_guide.rst [hunk 4]
```diff
@@ -1,219 +1,263 @@
...
+are pushed to the development branch on GitHub_.
 
-When bugs are discovered, they should be turned into test cases to
-prevent the bug from emerging again in the future [2]_.
+The following checks are performed with each pull request. The results
+of the checks are found near the end of the *Conversation* tab in each
+pull request. Most of these checks have been automated using `GitHub
+Actions`_. Checks that pass are marked with ✔️, while tests that fail
+are marked with ❌. Click on *Details* for information about why a
+particular check failed.
```
Comment (StanczakDominik) [inline docs/development/testing_guide.rst:77]:
```suggestion
particular check failed.

.. note::

   For first-time contributors, existing maintainers may need to manually
   enable your GitHub Action test runs. This is, believe it or not,
   indirectly caused by the invention of cryptocurrencies. Read more about this at
   https://docs.github.com/en/actions/managing-workflow-runs/approving-workflow-runs-from-public-forks
  
```

File: docs/development/testing_guide.rst [hunk 5]
```diff
@@ -350,104 +395,79 @@ functions or pass in tuples containing inputs and expected values.
...
+Pull requests must include tests of new or changed functionality before
+being merged.
 
-.. code-block:: python
+Best practices for writing tests
+--------------------------------
 
-  from numpy import sin, cos, pi
-  from plasmapy.utils.pytest_helpers import run_test
+The following guidelines are helpful suggestions for writing readable,
```
Comment (StanczakDominik) [inline docs/development/testing_guide.rst:409]: *[citation needed]* 😉

**Review by StanczakDominik — APPROVED** [review]
I know this is in draft, but I gave this a quick read-through. It's definitely better now, and if you wanted to merge it, I'd be happy to do so at any point.

File: docs/development/testing_guide.rst (hunk 4 above)
Comment (namurphy) [reply docs/development/testing_guide.rst:77]: Argh, I forgot about that, which is surprising given all the bitcoin mining I've been doing on GitHub pull requests to astropy_helpers.

**Comment by StanczakDominik** [comment]
Nick, please remember that this is not supposed to be an exhaustive primary source - unless I'm missing something! 😅

**Comment by namurphy** [comment]
> Nick, please remember that this is not supposed to be an exhaustive primary source - unless I'm missing something! sweat_smile

Yeah, you're probably right about that.

File: docs/development/testing_guide.rst [hunk 6]
```diff
@@ -1,219 +1,372 @@
...
-<https://help.github.com/articles/creating-a-pull-request/>`_ from your
-development branch to `PlasmaPy's GitHub repository`_.  The test suite will be run
-when the pull request is created and every time your development branch
-is subsequently updated.
-
-`Travis CI <https://travis-ci.org>`_ and `AppVeyor
-<https://www.appveyor.com>`_ run code tests and check that code examples
-in docstrings produce the expected output.  Travis CI runs the tests in
-a Linux/MacOS environment whereas AppVeyor runs the tests in a Windows
-environment.
```
Comment (namurphy) [inline docs/development/testing_guide.rst:70]: Travis CI and AppVeyor, I'll miss you! 😹

File: docs/development/testing_guide.rst [hunk 7]
```diff
@@ -1,219 +1,372 @@
...
+* `Python testing in PyCharm
+  <https://www.jetbrains.com/help/pycharm/testing-your-first-python-application.html>`__
+* `Python testing in Visual Studio Code
+  <https://code.visualstudio.com/docs/python/testing>`__
+* `Python testing in Atom <https://atom.io/packages/atom-python-test>`__
 
-Writing Tests
-=============
+Types of tests
+==============
```
Comment (namurphy) [inline docs/development/testing_guide.rst:260]: Is this section worth including here, or is it too general?

File: docs/development/testing_guide.rst [hunk 8]
```diff
@@ -350,222 +502,218 @@ functions or pass in tuples containing inputs and expected values.
...
-   David Bernstein writes: "Every bug exists because of a missing test
-   in a system.  The way to fix bugs using TDD [test-driven development]
-   is first write a failing test that represents the bug and then fix
-   the bug and watch the failing test turn green.
+Configurations for coverage tests are given in the ``[coverage:run]``
+and ``[coverage:report]`` sections of :file:`setup.cfg`. Codecov_
+configurations are given in :file:`.codecov.yaml`.
+
+Best practices
+==============
```
Comment (namurphy) [inline docs/development/testing_guide.rst:599]: Similarly, is this section worth including here, or should I make this available in a different way outside of our documentation?

**Comment by namurphy** [comment]
- [ ] Fix missing citation for `[Wilson2014]`

**Review by etjohnson — COMMENTED** [review]
I was able to set up, run and use both the purest and the tox tools. I like how it listed at which speed each one runs and the commands to alter are presented well in the code blocks.  With the environments each version should be installed in order to use py39, py38 etc, the section of types of tests was slightly confusing. The section writing tests was very useful, the way in which it walks through is useful and removes most of the ambiguity. May have to read over the fixtures and code coverage again however from reading over it seems I can get them to run. I like the idea of the best practices list at the end, this provides a useful recap and reference guide.

**Review by StanczakDominik — APPROVED** [review]
Looks good! A couple minor suggestions and ideas, but I'd be happy to have it merged as is.

File: docs/development/testing_guide.rst [hunk 9]
```diff
@@ -222,50 +375,51 @@ be included in the error message by using `f-strings
...
+.. caution::
+
+  Using ``==`` to compare floating point numbers can lead to brittle
+  tests because of slight differences due to limited precision, rounding
+  errors, and revisions to fundamental constants.
+
+Use `numpy.isclose` when comparing floating point numbers and
+`astropy.units.isclose` for |Quantity| instances to avoid these
+difficulties.  For comparing arrays, use `numpy.allclose`
+and `astropy.units.allclose` instead.
```
[the remark this answers is not in the document]
Comment (namurphy) [reply docs/development/testing_guide.rst:395]: That is a really helpful suggestion!

- [ ] Do this ↑

**Comment by namurphy** [comment]
pre-commit.ci autofix

**Comment by namurphy** [comment]
@etjohnson and @StanczakDominik — Thank you for the helpful reviews!  

My main changes were implementing most or all the changes that you suggested, revising the section on running different tox test environments, moving the definitions of unit & integration tests to the glossary page, and adding an image that shows the GitHub Actions checks.  We could expand the discussion on fixtures in another pull request if we'd like.

Thank you again!

**Pull request merged**